clippy_lints/src/decimal_bit_mask.rs

```diff
@@ -0,0 +1,106 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::source::SpanRangeExt;
+use rustc_data_structures::packed::Pu128;
+use rustc_hir::{AssignOpKind, BinOpKind, Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::Span;
+use rustc_span::source_map::Spanned;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for decimal literals used as bit masks in bitwise operations.
+    ///
+    /// ### Why is this bad?
+    /// Using decimal literals for bit masks can make the code less readable and obscure the intended bit pattern.
+    /// Binary or hexadecimal or octal literals make the bit pattern more explicit and easier to understand at a glance.
+    ///
+    /// ### Example
+    /// ```rust,no_run
+    /// let a = 14 & 6; // Bit pattern is not immediately clear
+    /// ```
+    /// Use instead:
+    /// ```rust,no_run
+    /// let a = 0b1110 & 0b0110;
+    /// ```
+    #[clippy::version = "1.87.0"]
+    pub DECIMAL_BIT_MASK,
+    nursery,
+    "use binary, hex or octal literals for bitwise operations"
+}
+
+declare_lint_pass!(DecimalBitMask => [DECIMAL_BIT_MASK]);
+
+fn check_bitwise_binary_expr(cx: &LateContext<'_>, e: &Expr<'_>) {
+    if let ExprKind::Binary(
+        Spanned {
+            node: BinOpKind::BitAnd | BinOpKind::BitOr | BinOpKind::BitXor,
+            ..
+        },
+        left,
+        right,
+    ) = &e.kind
+    {
+        for expr in [left, right] {
+            if let ExprKind::Lit(_) = &expr.kind
+                && !is_not_decimal_number(cx, expr.span)
+                && !is_power_of_twoish(expr)
+            {
+                emit_lint(cx, expr.span);
+            }
+        }
+    }
+}
+
+fn check_bitwise_assign_expr(cx: &LateContext<'_>, e: &Expr<'_>) {
+    if let ExprKind::AssignOp(
+        Spanned {
+            node: AssignOpKind::BitAndAssign | AssignOpKind::BitOrAssign | AssignOpKind::BitXorAssign,
+            ..
+        },
+        _,
+        expr @ Expr {
+            kind: ExprKind::Lit(lit),
+            ..
+        },
+    ) = &e.kind
```

Comment:
very much a nit: imo, these big patterns are too verbose -- what do you think about something like this?
```suggestion
    if let ExprKind::AssignOp(op, _, expr) = &e.kind
        && matches!(op.node, AssignOpKind::BitAndAssign | AssignOpKind::BitOrAssign | AssignOpKind::BitXorAssign)
        && let ExprKind::Lit(lit) = expr.kind
```